plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -127,3 +154,235 @@ counted out already so we should not immediately encounter those in `VBody`.
 Also, we currently reject `Constant` (has acceptable cost but not acceptable size).
 We may want to check their sizes instead of just rejecting them.
 -}
+
+-- | Inline fully applied functions. Note [Inlining of fully applied functions].
+callSiteInline
+    :: forall tyname name uni fun a m
+    . (ExternalConstraints tyname name uni fun m, Eq a, Ord name)
+    => InlineHints name a
+    -> PLC.BuiltinVersion fun
+    -> Term tyname name uni fun a
+    -> m (Term tyname name uni fun a)
+callSiteInline hints ver t = let
+        inlineInfo :: InlineInfo name fun a
+        inlineInfo = InlineInfo (snd deps) usgs hints ver
+        -- We actually just want the variable strictness information here!
+        deps :: (G.Graph Deps.Node, Map.Map PLC.Unique Strictness)
+        deps = Deps.runTermDeps ver t
+        usgs :: Usages.Usages
+        usgs = Usages.termUsages t
+    in liftQuote $ flip evalStateT mempty $ flip runReaderT inlineInfo $ processTerm t
+
+-- | Run the inliner on a `Core.Type.Term`.
+processTerm
+    :: forall tyname name uni fun a. (InliningConstraints tyname name uni fun, Eq a, Ord name)
+    => Term tyname name uni fun a -- ^ Term to be processed.
+    -> InlineM tyname name uni fun a (Term tyname name uni fun a)
+processTerm = \case
+    -- When we encounter a variable, consider inlining it.
+    v@(Var _a _n) -> do
+      considerInline v
+    Let a NonRec bs t -> do
+        -- Process non-recursive let bindings, checking for fully applied functions and
+        -- put them in the substitution. We do NOT remove the inlined binding here because it
+        -- may not be dead. We leave this to the dead code pass.
+        bs' <- traverse (processSingleBinding t) bs
+        t' <- processTerm t
+        pure $ Let a NonRec bs' t' -- we're not removing any bindings so it should be nonempty!
+    -- TODO Let a Rec bs t -> do
+      -- Process recursive let bindings, looking for let-bindings that are functions and are
+      -- recursively called.
+    t -> forMOf termSubterms t processTerm
+    -- | See note [Inlining of fully applied functions].
+    -- Consider inlining the variable that is NOT in the substitution by checking
+    -- (1) Is it fully applied?
+    -- (2) Are the cost and size acceptable?
+    where
+    considerInline ::
+        Term tyname name uni fun a -- the variable that is a function
+        -> InlineM tyname name uni fun a (Term tyname name uni fun a)
+    considerInline v@(Var a n) = do
+      -- look up the variable in the `CalledVar` map
+      varInfo <- gets (lookupCalled n)
+      case varInfo of
+        -- if it's not in the map, it's not a function, don't inline.
+        Nothing -> pure v
+        Just info -> do
+          let subst = calledVarDef info -- what we substitute in is its definition
+          isAcceptable <- acceptable subst
```

Comment:
I don't think this is correct.
E.g., `let f = \x y -> 3`, you should call `acceptable` on `3`, but you are calling it on `\x y -> 3`, as far as I can tell.